orttraining/orttraining/training_ops/cuda/communication/send.cc

```diff
@@ -28,145 +28,192 @@ ONNX_OPERATOR_KERNEL_EX(
-  // Begin of preparation for sending data. This time range includes
-  // the time for sending a scalar.
-  preRange.Begin();
-#endif
-
-  // Create buffers
-  const int tensor_num = static_cast<int>(element_types_.size());
-  // TODO move the following variables to member variables for extending life-time
-  // if we want to make the entire call async
-  std::vector<size_t> prefix_tensor_shape_sizes;
-  std::vector<int64_t> aggregated_tensor_shapes;
-  size_t aggregated_aligned_tensor_bytes = 0;
-  // tensor_offsets_in_bytes[i] is the starting byte of the i-th tensor in the send tensor buffer
-  std::vector<size_t> tensor_offsets_in_bytes;
-  // tensor_sizes_in_bytes[i] = (# of elements in the i-th tensor) * sizeof(the i-th tensor's element type)
-  std::vector<size_t> tensor_sizes_in_bytes;
-
-  // Same-rank communication is not allowed because we currently don't have async Send/Recv.
-  int world_rank;
-  MPI_Comm_rank(MPI_COMM_WORLD, &world_rank);
-  ORT_ENFORCE(world_rank != dst, "Sending data to rank ", dst, " on the rank ", world_rank, ".");
-
-  // Compute tensor shapes and sizes
-  size_t prefix_tensor_shape_size_sum = 0;
-  for (int i = 0; i < tensor_num; ++i) {
-    const Tensor* x_tensor = ctx->Input<Tensor>(i + 2);
-    prefix_tensor_shape_size_sum += x_tensor->Shape().NumDimensions();
-    prefix_tensor_shape_sizes.push_back(prefix_tensor_shape_size_sum);
-    aggregated_tensor_shapes.insert(aggregated_tensor_shapes.end(),
-                                    x_tensor->Shape().GetDims().begin(),
-                                    x_tensor->Shape().GetDims().end());
-
-    // Find the next aligned offset in the tensor buffer to meet alignment requirement
-    aggregated_aligned_tensor_bytes = GetAggregatedAlignedAddress(aggregated_aligned_tensor_bytes);
-    tensor_offsets_in_bytes.push_back(aggregated_aligned_tensor_bytes);
-    aggregated_aligned_tensor_bytes += x_tensor->SizeInBytes();
-    tensor_sizes_in_bytes.push_back(x_tensor->SizeInBytes());
-  }
-
-  IAllocatorUniquePtr<char> buffer = AllocateBufferOnCPUPinned<char>(
-      static_cast<size_t>(aggregated_aligned_tensor_bytes));
-
-  // Keep the sync copy in the previous design
-  // TODO they can be moved to async call after global stream becoming accessible
-  for (int i = 0; i < tensor_num; ++i) {
-    const Tensor* x_tensor = ctx->Input<Tensor>(i + 2);
-    ORT_ENFORCE(cudaMemcpy(buffer.get() + tensor_offsets_in_bytes[i], x_tensor->DataRaw(),
-                           tensor_sizes_in_bytes[i], cudaMemcpyDeviceToHost) == cudaSuccess);
-  }
-
-  // Prepare MPI communication info
-  int tensor_num_in_bytes = tensor_num * static_cast<int>(sizeof(size_t));
-  ORT_ENFORCE(tensor_num_in_bytes < INT_MAX,
+void Send::SendShapeInfo(
+    const int dst,
+    const int num_tensors,  // Number of sent tensors.
+    size_t aggregated_aligned_tensor_bytes,
+    std::vector<size_t> prefix_tensor_shape_sizes,
+    std::vector<int64_t> aggregated_tensor_shapes) const {
```

Comment:
As those parameter doesn't need to be changed, we can set them as const params,. Also, to avoid copy of vector, prefix_tensor_shape_sizes and aggregated_tensor_shapes can be passed by reference.

Reply:
same for SendData.

---
In reply to: [459817182](https://github.com/microsoft/onnxruntime/pull/4547#discussion_r459817182) [](ancestors = 459817182)

Reply:
Their pointers are stored as `void*` in
```
typedef struct {
  // Pointer to ONNX tensor's data on CPU.
  // It should be removed once we enable CUDA-aware MPI.
  void* buffer;
  // The size of buffer's content in bytes.
  int size;
  // Dst rank for Send and src rank for Recv.
  int rank;
  // Message's tag.
  int tag;
} CommInfo_t;
```
so they can't be constant. I haven't decided if we should keep that struct, so let's keep the non-constant for now. Those tensors are small, so coping them doesn't affect performance.

---
In reply to: [459817358](https://github.com/microsoft/onnxruntime/pull/4547#discussion_r459817358) [](ancestors = 459817358,459817182)